src/util/mpir_nodemap.h

```diff
@@ -322,13 +322,83 @@ static inline int MPIR_NODEMAP_populate_ids_from_mapping(char *mapping,
 
    Each process kvs_puts its hostname and stores the total number of
    processes (g_num_global).  Each process determines maximum node id
-   (g_max_node_id) and assigns a node id to each process (g_node_ids[]):
+   and assigns a node id to each process (g_node_ids[]):
 
      For each hostname the process seaches the list of unique nodes
      names (node_names[]) for a match.  If a match is found, the node id
      is recorded for that matching process.  Otherwise, the hostname is
      added to the list of node names.
 */
+
+/* TODO: make the fallback routine general, ie works across all PMI versions */
+#ifdef USE_PMI1_API
+static inline int MPIR_NODEMAP_build_nodemap_fallback(int sz, int myrank, int *out_nodemap,
+                                                      int *out_max_node_id)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    int key_max_sz = MPIR_pmi_max_key_size();
+    char *key = MPL_malloc(key_max_sz, MPL_MEM_OTHER);
+    char **node_names = MPL_malloc(sz * sizeof(char *), MPL_MEM_OTHER);
+    char *node_name_buf = MPL_malloc(sz * key_max_sz, MPL_MEM_OTHER);
+
+    for (int i = 0; i < sz; ++i) {
+        node_names[i] = &node_name_buf[i * key_max_sz];
+        node_names[i][0] = '\0';
+    }
+
+    mpi_errno = MPIR_NODEMAP_publish_node_id(sz, myrank);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    int max_node_id = -1;
+
+    for (int i = 0; i < sz; ++i) {
+        MPIR_Assert(max_node_id < sz);
+        if (i == myrank) {
+            /* This is us, no need to perform a get */
+            int ret;
+            char *hostname = (char *) MPL_malloc(sizeof(char) * MAX_HOSTNAME_LEN, MPL_MEM_ADDRESS);
+            ret = gethostname(hostname, MAX_HOSTNAME_LEN);
+            MPIR_ERR_CHKANDJUMP2(ret == -1, mpi_errno, MPI_ERR_OTHER, "**sock_gethost",
+                                 "**sock_gethost %s %d", MPIR_Strerror(errno), errno);
+            hostname[MAX_HOSTNAME_LEN - 1] = '\0';
```

Comment:
The total buffer size is `MAX_HOSTNAME_LEN`. `gethostname` should already terminated the string with `'\0'`. `hostname[MAX_HOSTNAME_LEN - 1] = '\0'` is a safeguard in case `gethostname` failed to do so (due to hitting max size).

Reply:
When `hostname` exceeds buffer size, we are not getting the full host name anyway (it doesn't fit), so we are making truncation.